From 04f7e4b71ce62dc5e3258ce5be9fa5d658493830 Mon Sep 17 00:00:00 2001
From: Quentin Casasnovas <quentinx.casasnovas@intel.com>
Date: Fri, 8 Feb 2013 15:54:58 +0100
Subject: [PORT FROM MAIN] wifi: Differentiate bcm chips firmware path

BZ: 83712

In the case of broadcom wireless chip, we need to know which chip is used
in order to send back the proper firmware path. This is necessary to add
support for dual chip.

Feature   : Wlan
Category  : Integration
Purpose   : ABSP-Adapter
Origin    : Upstream
Upstream-candidate: no

Change-Id: Icbab3813ec417b55fcf7572bb7b68570401081cd
Orig-Change-Id: I0bf65ea2515724eeded23a40b80b7c43cf7f8798
Signed-off-by: Quentin Casasnovas <quentinx.casasnovas@intel.com>
---
 wifi/Android.mk |   13 +++++++++++++
 wifi/wifi.c     |   44 +++++++++++++++++++++++++++++++++++++++++---
 2 files changed, 54 insertions(+), 3 deletions(-)

diff --git a/wifi/Android.mk b/wifi/Android.mk
index b4a6a7c..4596407 100644
--- a/wifi/Android.mk
+++ b/wifi/Android.mk
@@ -15,15 +15,28 @@ endif
 ifdef WIFI_FIRMWARE_LOADER
 LOCAL_CFLAGS += -DWIFI_FIRMWARE_LOADER=\"$(WIFI_FIRMWARE_LOADER)\"
 endif
+
 ifdef WIFI_DRIVER_FW_PATH_STA
 LOCAL_CFLAGS += -DWIFI_DRIVER_FW_PATH_STA=\"$(WIFI_DRIVER_FW_PATH_STA)\"
 endif
+LOCAL_CFLAGS += -DWIFI_DRIVER_43241_FW_PATH_STA=\"$(WIFI_DRIVER_43241_FW_PATH_STA)\"
+LOCAL_CFLAGS += -DWIFI_DRIVER_4334_FW_PATH_STA=\"$(WIFI_DRIVER_4334_FW_PATH_STA)\"
+LOCAL_CFLAGS += -DWIFI_DRIVER_4335_FW_PATH_STA=\"$(WIFI_DRIVER_4335_FW_PATH_STA)\"
+
 ifdef WIFI_DRIVER_FW_PATH_AP
 LOCAL_CFLAGS += -DWIFI_DRIVER_FW_PATH_AP=\"$(WIFI_DRIVER_FW_PATH_AP)\"
 endif
+LOCAL_CFLAGS += -DWIFI_DRIVER_43241_FW_PATH_AP=\"$(WIFI_DRIVER_43241_FW_PATH_AP)\"
+LOCAL_CFLAGS += -DWIFI_DRIVER_4334_FW_PATH_AP=\"$(WIFI_DRIVER_4334_FW_PATH_AP)\"
+LOCAL_CFLAGS += -DWIFI_DRIVER_4335_FW_PATH_AP=\"$(WIFI_DRIVER_4335_FW_PATH_AP)\"
+
 ifdef WIFI_DRIVER_FW_PATH_P2P
 LOCAL_CFLAGS += -DWIFI_DRIVER_FW_PATH_P2P=\"$(WIFI_DRIVER_FW_PATH_P2P)\"
 endif
+LOCAL_CFLAGS += -DWIFI_DRIVER_43241_FW_PATH_P2P=\"$(WIFI_DRIVER_43241_FW_PATH_P2P)\"
+LOCAL_CFLAGS += -DWIFI_DRIVER_4334_FW_PATH_P2P=\"$(WIFI_DRIVER_4334_FW_PATH_P2P)\"
+LOCAL_CFLAGS += -DWIFI_DRIVER_4335_FW_PATH_P2P=\"$(WIFI_DRIVER_4335_FW_PATH_P2P)\"
+
 ifdef WIFI_DRIVER_FW_PATH_PARAM
 LOCAL_CFLAGS += -DWIFI_DRIVER_FW_PATH_PARAM=\"$(WIFI_DRIVER_FW_PATH_PARAM)\"
 endif
diff --git a/wifi/wifi.c b/wifi/wifi.c
index 1d8ab33..32166c1 100644
--- a/wifi/wifi.c
+++ b/wifi/wifi.c
@@ -115,6 +115,12 @@ static const char P2P_CONFIG_FILE[]     = "/data/misc/wifi/p2p_supplicant.conf";
 static const char CONTROL_IFACE_PATH[]  = "/data/misc/wifi/sockets";
 static const char MODULE_FILE[]         = "/proc/modules";
 
+/*
+ * This gets defined by the script load_bcmdriver in
+ * vendor/intel/common/wifi/bcm_specific/
+ */
+static const char BCM_PROP_CHIP[]	= "wlan.bcm.chip";
+
 static const char SUPP_ENTROPY_FILE[]   = WIFI_ENTROPY_FILE;
 static unsigned char dummy_key[21] = { 0x02, 0x11, 0xbe, 0x33, 0x43, 0x35,
                                        0x68, 0x47, 0x84, 0x99, 0xa9, 0x2b,
@@ -938,14 +944,46 @@ int wifi_command(const char *ifname, const char *command, char *reply, size_t *r
 
 const char *wifi_get_fw_path(int fw_type)
 {
+    char bcm_prop_chip[PROPERTY_VALUE_MAX];
+
     switch (fw_type) {
     case WIFI_GET_FW_PATH_STA:
-        return WIFI_DRIVER_FW_PATH_STA;
+	if (property_get(BCM_PROP_CHIP, bcm_prop_chip, NULL)) {
+	    if (strstr(bcm_prop_chip, "43241"))
+		return WIFI_DRIVER_43241_FW_PATH_STA;
+	    else if (strstr(bcm_prop_chip, "4334"))
+		return WIFI_DRIVER_4334_FW_PATH_STA;
+	    else if (strstr(bcm_prop_chip, "4335"))
+		return WIFI_DRIVER_4335_FW_PATH_STA;
+	}
+	else
+	    return WIFI_DRIVER_FW_PATH_STA;
     case WIFI_GET_FW_PATH_AP:
-        return WIFI_DRIVER_FW_PATH_AP;
+	if (property_get(BCM_PROP_CHIP, bcm_prop_chip, NULL)) {
+	    if (strstr(bcm_prop_chip, "43241"))
+		return WIFI_DRIVER_43241_FW_PATH_AP;
+	    else if (strstr(bcm_prop_chip, "4334"))
+		return WIFI_DRIVER_4334_FW_PATH_AP;
+	    else if (strstr(bcm_prop_chip, "4335"))
+		return WIFI_DRIVER_4335_FW_PATH_AP;
+	}
+	else
+	    return WIFI_DRIVER_FW_PATH_AP;
     case WIFI_GET_FW_PATH_P2P:
-        return WIFI_DRIVER_FW_PATH_P2P;
+	if (property_get(BCM_PROP_CHIP, bcm_prop_chip, NULL)) {
+	    if (strstr(bcm_prop_chip, "43241"))
+		return WIFI_DRIVER_43241_FW_PATH_P2P;
+	    else if (strstr(bcm_prop_chip, "4334"))
+		return WIFI_DRIVER_4334_FW_PATH_P2P;
+	    else if (strstr(bcm_prop_chip, "4335"))
+		return WIFI_DRIVER_4335_FW_PATH_P2P;
+	}
+	else
+	    return WIFI_DRIVER_FW_PATH_P2P;
+    default:
+	    ALOGE("Unknown firmware type (%d)", fw_type);
     }
+
     return NULL;
 }
 
-- 
1.7.4.1

